File: 프로그래머스/LV2/코딩테스트 연습/완전탐색/피로도/피로도.py
ans = 0

def go(visit, dungeons, k, n):
    global ans
    if n > ans:
        ans = n

    for i in range(len(dungeons)):
        if not visit[i] and k >= dungeons[i][0]:
            visit[i] = True
            go(visit, dungeons, k - dungeons[i][1], n+1)
            visit[i] = False
    
def solution(k, dungeons):
    global ans
    visit = [False]*len(dungeons)
    go(visit, dungeons, k, 0)
    return ans

File: 프로그래머스/LV2/코딩테스트 연습/완전탐색/피로도/피로도.cpp
#include <string>
#include <vector>
using namespace std;

int ans = 0;

void go(vector<bool> visit, vector<vector<int>>& dungeons, int k, int n)
{
    if(n > ans)ans = n;

    for(int i = 0; i < dungeons.size(); i++)
    {
        if(!visit[i] && k >= dungeons[i][0])
        {
            visit[i] = true;
            go(visit, dungeons, k-dungeons[i][1], n+1);
            visit[i] = false;
        }
    }
}

int solution(int k, vector<vector<int>> dungeons) 
{
    vector<bool> visit;
    for(int i = 0; i < dungeons.size(); i++)visit.push_back(false);
    go(visit, dungeons, k, 0);
    return ans;
}